### PR TITLE
Feature/roc 1530 ccj amount

### PR DESCRIPTION
**Pull request by yarpo**
### JIRA link ###
https://tools.hmcts.net/jira/browse/ROC-1530

### Change description ###
Added new page to let claimant provide the answer if defendant has paid anything. If so - amount that was already paid.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/features/ccj/form/models/paidAmount.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      PaidAmountOption.NO
+    ]
+  }
+}
+
+export class PaidAmount implements Serializable <PaidAmount>, CompletableTask {
+
+  @IsDefined({ message: ValidationErrors.OPTION_REQUIRED })
+  @IsIn(PaidAmountOption.all(), { message: ValidationErrors.OPTION_REQUIRED })
+  option?: string
```
Comment (timja-kainos) [inline src/main/features/ccj/form/models/paidAmount.ts]: this should use a class, not a string

File: src/main/features/ccj/form/models/paidAmount.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import { CompletableTask } from 'models/task'
+
+export class ValidationErrors {
+  static readonly OPTION_REQUIRED: string = 'Choose option: yes or no'
+  static readonly AMOUNT_REQUIRED: string = 'Enter an amount'
+  static readonly AMOUNT_NOT_VALID: string = 'Invalid amount'
+}
+
+export class PaidAmountOption {
+  static readonly YES = 'yes'
```
Comment (timja-kainos) [inline src/main/features/ccj/form/models/paidAmount.ts]: This looks like a YesNoOption not specific to paying amount

File: src/main/features/ccj/routes/their-details.ts [hunk 3]
```diff
@@ -43,8 +43,9 @@ export default express.Router()
         if (form.hasErrors()) {
           renderView(form, res)
         } else {
+          const { externalId } = req.params
           user.ccjDraft.defendant.partyDetails = form.model
           await DraftCCJService.save(res, next)
-          res.redirect('todo')
+          res.redirect(Paths.paidAmountPage.uri.replace(':externalId', externalId))
```
Comment (timja-kainos) [inline src/main/features/ccj/routes/their-details.ts]: maybe leave this as todo? as this page isn't the next in the flow.
Coordinate with Ranjit to get this story merged first

File: src/main/features/ccj/views/paid-amount.njk [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+      {{ csrfProtection(csrf) }}
+
+      {% set error = form.errorFor('option') %}
+      {% set selection = form.valueFor('option') %}
+      <div class="form-group {% if error %} form-group-error {% endif %}">
+        <fieldset>
+          <legend class="visually-hidden">Has the defendant paid some of the amount owed?</legend>
+
+          {{
```
Comment (timja-kainos) [inline src/main/features/ccj/views/paid-amount.njk]: radioOption should never be used directly on a page, use radioGroup.
See https://github.com/hmcts/cmc-citizen-frontend/pull/16/files#diff-515e6f19e3ea26bc5e9ed26ab69f5eefR22

File: src/main/public/stylesheets/components/_form.scss [hunk 5]
```diff
@@ -16,3 +16,10 @@
     @include button($grey-3);
```
Comment (timja-kainos) [inline src/main/public/stylesheets/components/_form.scss]: can't comment on the image, but do we not already have it in the app for the amountBreakdown page?

File: src/test/features/ccj/routes/paid-amount.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+import { app } from '../../../../main/app'
+
+import * as idamServiceMock from '../../../http-mocks/idam'
+import * as claimStoreServiceMock from '../../../http-mocks/claim-store'
+import * as draftStoreServiceMock from '../../../http-mocks/draft-store'
+import { checkAuthorizationGuards } from './checks/authorization-check'
+import { PaidAmountOption } from 'ccj/form/models/paidAmount'
+
+const cookieName: string = config.get<string>('session.cookieName')
+const externalId = 'b17af4d2-273f-4999-9895-bce382fa24c8'
```
Comment (timja-kainos) [inline src/test/features/ccj/routes/paid-amount.ts]: externalId is available on the sampleClaimObj no need to hardcode, I see I messed it up in the their-details test can you fix?

File: src/test/features/ccj/routes/their-details.ts [hunk 7]
```diff
@@ -16,7 +16,9 @@ import { checkAuthorizationGuards } from './checks/authorization-check'
 
 const cookieName: string = config.get<string>('session.cookieName')
 
-const theirDetailsPage = Paths.theirDetailsPage.uri.replace(':externalId', 'b17af4d2-273f-4999-9895-bce382fa24c8')
+const externalId = 'b17af4d2-273f-4999-9895-bce382fa24c8'
```
Comment (timja-kainos) [inline src/test/features/ccj/routes/their-details.ts]: don't hardcode this see above

File: src/main/features/ccj/routes/their-details.ts (hunk 3 above)
Comment (yarpo) [inline src/main/features/ccj/routes/their-details.ts]: ok

File: src/main/features/ccj/form/models/paidAmount.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  @IsPositive({ message: ValidationErrors.AMOUNT_NOT_VALID })
+  amount?: number
+
+  constructor (option?: PaidAmountOption, amount?: number) {
+    this.option = option
+    this.amount = amount
+  }
+
+  static fromObject (value?: any): PaidAmount {
+    console.log(value)
```
Comment (timja-kainos) [inline src/main/features/ccj/form/models/paidAmount.ts]: console.log

File: src/main/features/ccj/form/models/paidAmount.ts (hunk 1 above)
Comment (yarpo) [inline src/main/features/ccj/form/models/paidAmount.ts]: done

File: src/main/features/ccj/form/models/paidAmount.ts (hunk 2 above)
Comment (yarpo) [inline src/main/features/ccj/form/models/paidAmount.ts]: I'll add it

File: src/main/features/ccj/views/paid-amount.njk (hunk 4 above)
Comment (yarpo) [inline src/main/features/ccj/views/paid-amount.njk]: Please notice that in here we need to be able to add additional code for expanded form when option is selected. Can we do it with radioGroup? If so - how?

Comment (timja-kainos) [inline src/main/features/ccj/views/paid-amount.njk]: that functionality is built into the option macro which is in the chain of:
radioGroup => radioOption => option
so it shouldn't be any different

File: src/test/features/ccj/routes/paid-amount.ts (hunk 6 above)
Comment (yarpo) [inline src/test/features/ccj/routes/paid-amount.ts]: ok

File: src/main/public/stylesheets/components/_form.scss (hunk 5 above)
Comment (yarpo) [reply src/main/public/stylesheets/components/_form.scss]: we have two different approaches:

```.input-symbol-pound::before {
    content: '£';
    left: 7px;
    position: absolute;
    top: 1px;
}```

and this image.

I'll talk to UX guys to find out if this is what we want. If not - I'll align it.

File: src/test/features/ccj/routes/paid-amount.ts (hunk 6 above)
Comment (yarpo) [reply src/test/features/ccj/routes/paid-amount.ts]: done

File: src/test/features/ccj/routes/their-details.ts (hunk 7 above)
Comment (yarpo) [reply src/test/features/ccj/routes/their-details.ts]: done

File: src/main/features/ccj/views/paid-amount-summary.njk [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        </tr>
+        {% if interestDetails %}
+        <tr>
+          <th scope="row">
+            <details role="group">
+              <summary role="button" aria-controls="details-content-0" aria-expanded="false">
+                {{ t("Interest to date of judgment") }}
+              </summary>
+              <div class="panel text-small" id="details-content-0" aria-hidden>
+                {{ t("Interest calculated at %s&#37; for %s days", {
```
Comment (geek-soft) [inline src/main/features/ccj/views/paid-amount-summary.njk]: A `"Interest calculated at %s%%; for %s days"` could be more readable and not require `| safe`.

File: src/main/features/ccj/views/paid-amount.njk [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+    <form method="post" novalidate>
+
+      {{ csrfProtection(csrf) }}
+
+      {% set error = form.errorFor('option') %}
+      {% set selection = form.valueFor('option').value %}
+      <div class="form-group {% if error %} form-group-error {% endif %}">
+        <fieldset>
+          <legend class="visually-hidden">Has the defendant paid some of the amount owed?</legend>
```
Comment (geek-soft) [inline src/main/features/ccj/views/paid-amount.njk]: That needs to be translated as well.

File: src/main/public/stylesheets/components/_form.scss (hunk 5 above)
Comment (geek-soft) [reply src/main/public/stylesheets/components/_form.scss]: Lets not use an image, can't see any reason why image would be better then single pound char.

File: src/test/a11y/a11y.ts [hunk 11]
```diff
@@ -73,5 +73,4 @@ describe('Accessibility', () => {
   checkPaths(ClaimIssueErrorPaths)
   checkPaths(DefendantFirstContactPaths)
   checkPaths(DefendantFirstContactErrorPaths)
-  checkPaths(DefendantResponsePaths)
```
Comment (geek-soft) [inline src/test/a11y/a11y.ts]: Why has it been removed? Response journey needs to be covered by a11y tests.

File: src/test/features/ccj/routes/paid-amount-summary.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+  describe('on GET', () => {
+    checkAuthorizationGuards(app, 'get', paidAmountSummaryPage)
+
+    context('when user authorised', () => {
+      beforeEach(() => {
+        idamServiceMock.resolveRetrieveUserFor(1, 'cmc-private-beta')
+      })
+
+      it('should return 500 and render error page when cannot retrieve claims', async () => {
```
Comment (geek-soft) [inline src/test/features/ccj/routes/paid-amount-summary.ts]: I would be also interested to see the behaviour when:
- retrieving a CCJ draft failed
- that page is not accessible when claim is not eligible for CCJ e.g. `checkAlreadySubmittedGuard(app, 'get', ResponsePaths.defendantAddressPage.uri)` done for response

File: src/test/features/ccj/routes/paid-amount.ts [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+        await request(app)
+          .get(paidAmountPage)
+          .set('Cookie', `${cookieName}=ABC`)
+          .expect(res => expect(res).to.be.successful.withText('Has the defendant paid some of the amount owed?'))
+      })
+    })
+    describe('on POST', () => {
+      checkAuthorizationGuards(app, 'post', paidAmountPage)
+      context('when user authorised', () => {
```
Comment (geek-soft) [inline src/test/features/ccj/routes/paid-amount.ts]: Would be great to give these lines to breath.

File: src/test/features/ccj/routes/paid-amount.ts [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+            draftStoreServiceMock.resolveSave('ccj')
+
+            await request(app)
+              .post(paidAmountPage)
+              .set('Cookie', `${cookieName}=ABC`)
+              .send(validFormData)
+              .expect(res => expect(res).to.be.redirect.toLocation(paidAmountSummaryPage))
+          })
+        })
+        context('when form is invalid', async () => {
```
Comment (geek-soft) [inline src/test/features/ccj/routes/paid-amount.ts]: Please prefix this block with blank line here as well.

File: src/test/features/ccj/routes/paid-amount.ts [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+  describe('on GET', () => {
+    checkAuthorizationGuards(app, 'get', paidAmountPage)
+
+    context('when user authorised', () => {
+      beforeEach(() => {
+        idamServiceMock.resolveRetrieveUserFor(1, 'cmc-private-beta')
+      })
+
+      it('should return 500 and render error page when cannot retrieve claims', async () => {
```
Comment (geek-soft) [inline src/test/features/ccj/routes/paid-amount.ts]: The same here, it would be good to see the behaviour when:

- retrieving a CCJ draft failed
- that page is not accessible when claim is not eligible for CCJ

File: src/test/a11y/a11y.ts (hunk 11 above)
Comment (yarpo) [reply src/test/a11y/a11y.ts]: accidentally. I'll revert it

File: src/main/features/ccj/form/models/paidAmount.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+export class PaidAmount implements Serializable <PaidAmount> {
+
+  @IsDefined({ message: ValidationErrors.OPTION_REQUIRED })
+  @IsIn(PaidAmountOption.all(), { message: ValidationErrors.OPTION_REQUIRED })
+  option?: PaidAmountOption
+
+  @ValidateIf(o => o.option === PaidAmountOption.YES)
+  @IsDefined({ message: ValidationErrors.AMOUNT_REQUIRED })
+  @IsPositive({ message: ValidationErrors.AMOUNT_NOT_VALID })
+  amount?: number
```
[the remark this answers is not in the document]
Comment (timja-kainos) [reply src/main/features/ccj/form/models/paidAmount.ts]: how would that work for a form model?

File: src/test/features/ccj/routes/paid-amount-summary.ts (hunk 12 above)
Comment (timja-kainos) [reply src/test/features/ccj/routes/paid-amount-summary.ts]: there's a test for the guard specifically, I didn't want to go running that test on every view, can do if needed but it seemed overkill..

File: src/main/features/ccj/form/models/paidAmount.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (geek-soft) [reply src/main/features/ccj/form/models/paidAmount.ts]: @yarpo ignore my comment. Did not see the `deserialize` method which mutates the instance state.

When I posted my comment I was only thinking of form submission and `fromObject` method.

File: src/main/features/ccj/views/paid-amount-summary.njk [hunk 17]
```diff
@@ -0,0 +1,95 @@
+{% extends "layout.njk" %}
+{% set title = 'Money claim' %}
+{% from "includes/macros/form.njk" import submitButton, csrfProtection %}
+
+{% block content %}
+<div class="grid-row">
```
Comment (timja-kainos) [inline src/main/features/ccj/views/paid-amount-summary.njk]: can you attach a screenshot please?

File: src/main/features/ccj/views/paid-amount-summary.njk [hunk 18]
```diff
@@ -0,0 +1,95 @@
+{% extends "layout.njk" %}
+{% set title = 'Money claim' %}
+{% from "includes/macros/form.njk" import submitButton, csrfProtection %}
```
Comment (timja-kainos) [inline src/main/features/ccj/views/paid-amount-summary.njk]: no csrfProtection needed if there isn't a form

File: src/test/features/ccj/routes/paid-amount-summary.ts (hunk 12 above)
Comment (geek-soft) [reply src/test/features/ccj/routes/paid-amount-summary.ts]: @timja-kainos can you point me to the guard test? It might be enough if it makes sure that guard is applied to each CCJ path. If it is not we might miss the issue where guard has not been properly applied to some CCJ paths.

File: yarn.lock [hunk 19]
```diff
@@ -112,9 +112,9 @@
   version "8.0.24"
```
Comment (geek-soft) [inline yarn.lock]: Can we revert changes in this file? These are not related to the PR I think.

File: src/test/features/ccj/routes/paid-amount-summary.ts (hunk 12 above)
Comment (timja-kainos) [reply src/test/features/ccj/routes/paid-amount-summary.ts]: https://github.com/hmcts/cmc-citizen-frontend/blob/master/src/test/features/ccj/routes/ccjGuard.ts

File: src/main/features/ccj/form/models/paidAmount.ts [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+  static readonly OPTION_REQUIRED: string = 'Choose option: yes or no'
+  static readonly AMOUNT_REQUIRED: string = 'Enter an amount'
+  static readonly AMOUNT_NOT_VALID: string = 'Invalid amount'
+}
+
+export class PaidAmount implements Serializable <PaidAmount> {
+
+  @IsDefined({ message: ValidationErrors.OPTION_REQUIRED })
+  @IsIn(PaidAmountOption.all(), { message: ValidationErrors.OPTION_REQUIRED })
+  option?: PaidAmountOption
```
[the remark this answers is not in the document]
Comment (yarpo) [reply src/main/features/ccj/form/models/paidAmount.ts]: if we add readonly here `deserialise(input)` doesn't work.

File: src/main/features/ccj/form/models/paidAmount.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (yarpo) [reply src/main/features/ccj/form/models/paidAmount.ts]: if we add readonly here `deserialise(input)` doesn't work.

File: src/main/features/ccj/views/paid-amount-summary.njk (hunk 9 above)
Comment (yarpo) [reply src/main/features/ccj/views/paid-amount-summary.njk]: ok

File: src/main/features/ccj/views/paid-amount.njk (hunk 10 above)
Comment (yarpo) [reply src/main/features/ccj/views/paid-amount.njk]: done

File: src/test/features/ccj/routes/paid-amount.ts (hunk 13 above)
Comment (yarpo) [reply src/test/features/ccj/routes/paid-amount.ts]: ok

File: src/test/features/ccj/routes/paid-amount.ts (hunk 14 above)
Comment (yarpo) [reply src/test/features/ccj/routes/paid-amount.ts]: ok

File: src/main/features/ccj/views/paid-amount-summary.njk (hunk 18 above)
Comment (yarpo) [reply src/main/features/ccj/views/paid-amount-summary.njk]: removed

File: yarn.lock (hunk 19 above)
Comment (yarpo) [reply yarn.lock]: ok

File: src/main/public/stylesheets/components/_form.scss (hunk 5 above)
Comment (yarpo) [reply src/main/public/stylesheets/components/_form.scss]: I have talked to Tarik and we should use css solution, not background img. Updated.

File: src/test/features/ccj/routes/paid-amount.ts (hunk 15 above)
Comment (yarpo) [reply src/test/features/ccj/routes/paid-amount.ts]: Added test for "retrieving a CCJ draft failed"

File: src/test/features/ccj/routes/paid-amount-summary.ts (hunk 12 above)
Comment (yarpo) [reply src/test/features/ccj/routes/paid-amount-summary.ts]: added test case for "retrieving a CCJ draft failed"

Comment (geek-soft) [reply src/test/features/ccj/routes/paid-amount-summary.ts]: I think having guard tests in separate file is fine but it is testing just one particular route at the moment. I would prefer to have tests for all eligible routes. Maybe could be done by iterating through list of routes in `ccj/paths`

File: src/main/features/ccj/views/paid-amount-summary.njk [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+        <tr>
+          <th scope="row">
+            <details role="group">
+              <summary role="button" aria-controls="details-content-0" aria-expanded="false">
+                {{ t("Interest to date of judgment") }}
+              </summary>
+              <div class="panel text-small" id="details-content-0" aria-hidden>
+                {{ t("Interest calculated at %s%% for %s days", {
+                  postProcess: 'sprintf', sprintf: [ interestDetails.rate, interestDetails.numberOfDays ]
+                  }) | safe
```
Comment (geek-soft) [inline src/main/features/ccj/views/paid-amount-summary.njk]: We probably do not need to use `| safe` pipe anymore

File: src/main/views/includes/macros/form.njk [hunk 22]
```diff
@@ -19,6 +19,27 @@
   {{ input(label, name, form, hint, 'text', inputClass, autocomplete) }}
 {% endmacro %}
 
+{% macro poundInput(label, name, form, hint = '') %}
+  {% set error = form.errorFor(name) %}
```
Comment (geek-soft) [inline src/main/views/includes/macros/form.njk]: Would be possible to use `input` here? I know there are some differences. Maybe we could make `input` macro more configurable to introduce things like:
- max length
- length
- wrapper class (for `<span class="input-symbol-pound">`.

File: src/test/features/ccj/routes/paid-amount.ts [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+    })
+
+    describe('on POST', () => {
+      checkAuthorizationGuards(app, 'post', paidAmountPage)
+
+      context('when user authorised', () => {
+        beforeEach(() => {
+          idamServiceMock.resolveRetrieveUserFor(1, 'cmc-private-beta')
+        })
+
```
Comment (geek-soft) [inline src/test/features/ccj/routes/paid-amount.ts]: We should also have test case for `should return 500 and render error page when cannot retrieve CCJ draft` for `POST` requests.

File: src/main/views/includes/macros/form.njk (hunk 22 above)
Comment (timja-kainos) [reply src/main/views/includes/macros/form.njk]: I think it can already do some of that, but having a specific macro for this is better IMO
Its reusable and it configured correctly

File: src/test/features/ccj/routes/paid-amount-summary.ts (hunk 12 above)
Comment (yarpo) [reply src/test/features/ccj/routes/paid-amount-summary.ts]: added tests for guard src/test/features/ccj/routes/ccjGuard.ts

File: src/main/views/includes/macros/form.njk (hunk 22 above)
Comment (yarpo) [reply src/main/views/includes/macros/form.njk]: I have talked to Tarik and "pound input" will be used in many places in the future.

I'm not sure if there is any gain of adding more arguments to input / any other macro and make implementation of it so much more complex. Maintaining of this wouldn't be easier.

To be honest I'd rather create a ticket for GOV template so that pound input would just have css class on input, not the whole structure of form implemented differently.

File: src/test/features/ccj/routes/paid-amount.ts (hunk 23 above)
Comment (yarpo) [reply src/test/features/ccj/routes/paid-amount.ts]: done
